src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -0,0 +1,149 @@
+#include "memory/allocation.hpp"
+#include "memory/arena.hpp"
+#include "utilities/growableArray.hpp"
+#include "runtime/os.hpp"
+
+// A free list, growth only, allocator for a specific type E.
+// The allocator returns 'pointers' of 4-bytes in size, allowing for
+// memory savings if a pointer-heavy self-referential structure is used.
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  // Make the index opaque.
+  struct I {
+    int32_t idx;
+    bool operator !=(I other) {
+      return idx != other.idx;
+    }
+    bool operator==(I other) {
+      return idx == other.idx;
+    }
+  };
```

Comment:
Now we use access specifiers to deny access to the index, except for the allocator, which is a friend. We do want the index to be opaque, so we do this. On top of that, we use an `_owner` field in debug mode to assert that we don't free to the wrong allocator.

Reply:
> Now we use access specifiers to deny access to the index, except for the allocator, which is a friend. We do want the index to be opaque, so we do this. On top of that, we use an `_owner` field in debug mode to assert that we don't free to the wrong allocator.

The last part I don't like. It causes `I` to quadruple in size in debug builds only. For something that is potentially used as member in bunch of different data structures, it can cause considerable deltas in memory layout between debug and release builds. I don't think that is a good thing. What you test in debug should be close to what's running at a customer.

I also don't think it is particularly useful. I would instead do a simple range check on I. That gives you a part of the sanity checks without the negatives. Couple that with a sanity check wrt to slot state (don't free free slots) you are pretty well covered already.

Other than that, I think using an own class to hide a simple i32 is unnecessary complex, but I leave that up to you. If you do this, please take care of constness. And make it work for its money, e.g. by giving it an is_nil instead of exposing nil, and requiring the user to manually compare it.

Reply:
Alright, I get the counterpoint re: memory usage of the slots. I can live with that, though I am a bit disappointed as I had double-free coverage in mind also :).

>Other than that, I think using an own class to hide a simple i32 is unnecessary complex, but I leave that up to you. If you do this, please take care of constness. And make it work for its money, e.g. by giving it an is_nil instead of exposing nil, and requiring the user to manually compare it.

Yes, I'll make `_idx` const. `nil` is also exposed so that you can create nil pointers, which is important. I wanted to make the `I(int32_t idx)` constructor private, as only the allocator should construct pointers. This turned out to not be possible due to the design of `GrowableArray`.